src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,506 @@
+﻿/* Copyright 2020–present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using MongoDB.Bson;
+using MongoDB.Bson.Serialization;
+using MongoDB.Driver.Core.Connections;
+using MongoDB.Driver.Core.Misc;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Security;
+using System.Threading.Tasks;
+
+namespace MongoDB.Driver.Core.Authentication
+{
+    /// <summary>
+    /// The Mongo AWS authenticator.
+    /// </summary>
+    public class MongoAWSAuthenticator : SaslAuthenticator
+    {
+        // constants
+        private const int ClientNonceLength = 32;
+        private const string AccessKeyIdExceptionMessage = "A MONGODB-AWS must have access key ID.";
```

Comment:
Thanks for taking my suggestion regarding the constants! 
I think my other comment also applies here:
I think in order for this sentence to be grammatically correct (and following the pattern used for PLAIN credentials), this might be better phrased as:
"A MONGODB-AWS **credential** must have an access key id."
"A MONGODB-AWS **credential** must have a secret acess key."

I suppose theoretically we could make these constants internal so that MongoCredential.cs could use it as well, but I leave that up to you :)

Reply:
consts order?